torch/onnx/_internal/fx/dynamo_graph_extractor.py

```diff
@@ -0,0 +1,209 @@
+from __future__ import annotations
+
+import copy
+from typing import Any, Callable, Mapping, Optional, Sequence, Tuple, Union
+
+import torch.onnx
+from torch.onnx import utils as onnx_utils
+from torch.onnx._internal import exporter
+from torch.onnx._internal.fx import io_adapter as io_adapter
+
+
+class DynamoOptimizeGraphCaptureCompiler:
+    def __init__(self):
+        self.captured_graph: Optional["torch.fx.GraphModule"] = None
+        self.captured_graph_count = 0
+
+    def __call__(self, graph_module: "torch.fx.GraphModule", _):
+        assert self.captured_graph_count == 0
+        self.captured_graph = graph_module
+        self.captured_graph_count += 1
+        return graph_module
+
+
+class DynamoOptimize(exporter.FXGraphExtractor):
+    """Generates a FX GraphModule using torch.dynamo.optimize API
+    Args:
+        backend: Specify which backend compiler must be used by Torch Dynamo.
```

Comment:
Why not? It wraps all `torch._dynamo.optimize` flags, which we could use

Today we are setting the options we want as default in this class, but we could create a new inherited type that calls the ctor with a different flag and verify its behavior

Reply:
Because `DynamoOptimize`'s expectation is different.

* `DynamoOptimize`'s main logic requires this backend to capture fx graph and passing it out through a variable. 
* `torch._dynamo.optimize`'s logic is to do backend custom optimization and return optimized **python callable**. The api doesn't expose the inner fx graph at all. And for some backends, there may not be an fx graph anymore.

These are not revealed in the docstring. Providing any other backend mentioned in the docstring breaks `DynamoOptimize`.

Reply:
> Because `DynamoOptimize`'s expectation is different.
> 
> * `DynamoOptimize`'s main logic requires this backend to capture fx graph and passing it out through a variable.
> * `torch._dynamo.optimize`'s logic is to do backend custom optimization and return optimized **python callable**. The api doesn't expose the inner fx graph at all. And for some backends, there may not be an fx graph anymore.
> 
> These are not revealed in the docstring. Providing any other backend mentioned in the docstring breaks `DynamoOptimize`.

Regarding `torch._dynamo.optimize`, the limitations above are a possible (undesired) implementation detail. We also can have a different implementation in which `torch._dynamo.optimize` guarantees that there will always be a graph module - next we use a passthrough custom backend that simply returns the traced GraphModule. That is enough to leverage all good stuff Meta is doing with `torch.optimize`s FX tracing to support all models they see around on github. 

The export is not bound to use any backend that would do anything weird with the always-existing FX graph. Does it make sense? 

In fact, this is the way we were leveraging `torch._dynamo.optimize` from the beginning.

Reply:
You lost me completely lol. There is a way out of the problem but I'm not sure if that's what you meant. Unless dealt with, the problem `Providing any other backend mentioned in the docstring breaks DynamoOptimize.` remains, either the docstring or code has to change. Let's discuss this in the call this afternoon.